### ISSUE TITLE
Crashing on Windows 10

### ISSUE THREAD

**Issue by bukurat** [jziolkowski/tdm#10]
Python 3.7 on Windows 10.
After configuring MQTT host, some Sonoff devices found then program crashes.
Program now opens, then crashes after a couple of seconds.

Console:

D:\python\tdm>python tdm.py
Traceback (most recent call last):
  File "tdm.py", line 337, in mqtt_message
    self.device_model.updateValue(found.index, DevMdl.RESTART_REASON, payload['RestartReason'])
KeyError: 'RestartReason'

**Comment by jziolkowski** [comment]
It seems that one of your devices has older firmware, which doesn't report the restart reason. I'll push a fix in a few mins.

**Comment by jziolkowski** [comment]
Fix pushed.

**Comment by bukurat** [comment]
I Tried the latest build:

PS D:\python\tdm-master> python tdm.py
Traceback (most recent call last):
  File "D:\python\tdm-master\Util\models.py", line 102, in data
    if val.startswith("0T"):
AttributeError: 'int' object has no attribute 'startswith'
PS D:\python\tdm-master>

Error Log:
2019-03-01 12:07:25	stat/sonoff7/POWER	OFF	Expecting value
2019-03-01 12:07:25	stat/Bar_Light/POWER	OFF	Expecting value
2019-03-01 12:07:25	stat/sonoff01/POWER	ON	Expecting value
2019-03-01 12:07:25	stat/sonoff4/POWER	OFF	Expecting value
2019-03-01 12:07:25	stat/sonoffrgb1/POWER	OFF	Expecting value
2019-03-01 12:07:25	stat/sonoff3/POWER	OFF	Expecting value
2019-03-01 12:07:25	stat/ailight06/POWER	OFF	Expecting value

Devices from cfg file
[Devices]
sonoff4\full_topic=%prefix%/%topic%/
sonoff4\friendly_name=S
sonoffs20-2\full_topic=%prefix%/%topic%/
sonoffs20-2\friendly_name=S
sonoffs20-1\full_topic=%prefix%/%topic%/
sonoffs20-1\friendly_name=S
sonoff01\full_topic=%prefix%/%topic%/
sonoff01\friendly_name=S
ailight06\full_topic=%prefix%/%topic%/
ailight06\friendly_name=A
sonoff3\full_topic=%prefix%/%topic%/
sonoff3\friendly_name=%prefix%/%topic%/
Bar_Light\full_topic=%prefix%/%topic%/
Bar_Light\friendly_name=S
sonoffrgb1\full_topic=%prefix%/%topic%/
sonoffrgb1\friendly_name=%prefix%/%topic%/
sonoff7\full_topic=%prefix%/%topic%/
sonoff7\friendly_name=%prefix%/%topic%/

**Comment by jziolkowski** [comment]
I'm pushing the fix to the erroneous POWER log message.
Now, for the <pre>val.startswith("0T")</pre> I need status 0 for this device

**Comment by jziolkowski** [comment]
try the latest commit, I've added a str cast for device uptime in case it's malformed/nonexistent

**Comment by bukurat** [comment]
Getting there. TDM opens and shows data, however its somewhat different to the image in the wiki.
My name column shows a single character rather than the  friendly name.
If I double click on  row the details window shows for a second or so, the program then closes  with this traceback:
PS D:\python\tdm-master> python tdm.py
Traceback (most recent call last):
  File "D:\python\tdm-master\GUI\DeviceConfig.py", line 182, in mqtt_message
    self.program_model.item(6+i, 1).setText(msg['FriendlyName'][i])
AttributeError: 'NoneType' object has no attribute 'setText'

I have attached a file containing the output from status 0 on the two firmware versions currently in use.

[Status 0.txt](https://github.com/jziolkowski/tdm/files/2921639/Status.0.txt)

**Comment by jziolkowski** [comment]
Based on the file attached, you're using very old firmware, which is no longer supported.
The bug is because recent firmware report friendly name as a 4-element array, instead of string. I'll make a workaround for that, but you really should upgrade :)

**Comment by bukurat** [comment]
Thanks. I usually subscribe to the "If its not broken, don't fix it" school of thought and everything works to my satisfaction. Having said that, I've synced my files with the Tasmota repository and will compile a later version of the firmware.
 
Does TDM have an inbuilt web server for OTA upgrades?

**Comment by jziolkowski** [comment]
I agree completely, but 6.3 and 6.4 branches added lots of nice things, plus stability improvements. Definitely worth trying on a device or two for starters.

Built-in OTA server is in the works :)

**Comment by jziolkowski** [comment]
latest commit includes workaround for that FriendlyName issue

**Comment by bukurat** [comment]
I took your advice and compiled 6.4.1 classic, which is now installed on 11 of my devices.
I still have 3 devices on 5.12. They are reporting 256k free memory so I'll have to pull them out and re-flash them over the serial connection. Interestingly they are all compiled with the 2.4.1 core.

I had a couple with the same build and reporting the same free memory in the spares box. They re-flashed over the serial link and work correctly.

I notice that your details window doesn't show the Program Size, although it does show the Free Program Space and Free Memory.

I noticed that when I double-click on a row to see the information the console reports a single 
{'Command': 'Unknown'} 
If I switch to the 'Rules and Timers' tab the console responds with an additional 14 lines of {'Command': 'Unknown'}

Switching back and forth between the three tabs results in the program entering a non responsive mode.
If I force the program to close  nothing further is output to the console.

**Comment by jziolkowski** [comment]
Program Size is not visible due to not being exported via MQTT. I'll ask Theo to add this value in some future version.

About rules and timers: did you include rules and timers in your compiled images?

**Comment by Jason2866** [comment]
What would be the added value of knowing Program Size?
You cant update a device, if memory model differs from old to new version. No matter of program size.
With function OTA magic it is not relevant how big is old or new version.

**Comment by jziolkowski** [comment]
@Jason2866 only because I want the info tab to mimic the information page from webui. Let's call it aesthetics. Tho I doubt if that's reason enough for Theo ;)

**Comment by bukurat** [comment]
Rules and timers. Probably not in the classic. I have timers in some  POW devices I flashed today. I’ll check them later.
Command unknown. I’m not sure which rows were being accessed  the text appeared in the windows powershell I opened to run TDM.

@Jason2866  I had some devices that OTA magic couldn’t save as the available program space was less than the that of the minimum build. There is a chance that there could be others in the world with the same issue.

**Comment by andrethomas** [comment]
I think TDM opens the possibility of using a Tasmota device without the need of the web UI... I mean intentionally compiled to NOT include the WebUI in favor of the flash space being used by other functionality to still achieve a binary footprint of lower than 502k allowing the intermediate upgrade step to minimal to be excluded during the upgrade process.

So the more it mimics the behavior of the Web UI the more likely it is that people will no longer need the web ui... at least being able to exclude it by choice during self compiling using "TFB (c) (tm) @jziolkowski " ;)

**Comment by jziolkowski** [comment]
Well indeed that's possible. I wonder what's the code difference between removing everything HTTP related vs keeping only REST API without the GUI in Tasmota.

Also, I'd love to see which command was unknown in the JSON response. Would make it much easier to debug things.

**Comment by bukurat** [comment]
@jziolkowski I’ll see if I can isolate the “Command unknown” by watching the DUTs console when double clicking on its entry. Now that I have a build with timers it may be different.

**Comment by jziolkowski** [comment]
Good idea. I'll paste you the list of commands sent from TDM to the device so you don't have to guess.

Ok, upon opening the device config tab these are initial commands:
```
status 0
module
modules
gpio
gpios
timers
```
when you switch to tab "Rules" it issues 
```
rule1
timers
timer1
ruletimer
var1 up to var5
mem1 up to mem5
```

**Comment by bukurat** [comment]
I made this file before reading your last message so I'll let you look at it.  3 different compiles and the results.
I've marked the lines  where Command Unknown is with a bunch of ******* to help them stand out and give you something to search for.

I also noticed that if I type crap at the console window  - a way of delimiting when there is no clear - and the device row s selected in TDM, TDM will become unresponsive.
